README.md
# UT crypto proxy

Proxy for MLE and CBC

Listen on localhost and proxy requests to a remote URI.
It will encrypt JSON RPC requests and and decrypt responses.

## Project links

- [Continuous Integration (Jenkins)](https://jenkins.softwaregroup.com/view/master/job/ut/job/ut-babelfish/)
- [Static Code Analysis (SonarQube)](https://sonar.softwaregroup.com/dashboard?id=ut-babelfish%3Aorigin%2Fmaster)

## Usage

```npx ut-babelfish --proxy.uri=http://hostname:8004```

## Configuration

This module uses [ut-config](https://github.com/softwaregroup-bg/ut-config),
so it can be configured using command line, rc file, etc.

For example it will look for a file named `.ut_babelfish_devrc` in the
locations mentioned in `ut-config` readme.

Example YAML configuration, with the defaults:

```yaml
log:
    level: info
server:
    host: localhost
    port: 8091
    tls: false
proxy:
    passThrough: true
    xforward: false
    uri: http://localhost:8004
    tls: false
mle:
    routes:
        - /rpc/{path*}
    sign:
        kty: EC
        d: zmnSC_P5Xzefte7vkdINXLAN2LeBgC0S5QTcPO2mI5vo62chc_zHAYhcobGPQGNJ
        use: sig
        crv: P-384
        x: G4JWlybVRkliYWLLFdXDj0CjMjnkXeyiunzQswR3izK-jxvMIYdjVB52Rty5yZN9
        y: JndKKF7RQf97idkaLPLsv_jkZPBw-MJFogDqri87vvnpAEf1qyHnQTmK_gAhLAgo
        alg: ES384
    encrypt:
        kty: EC
        d: Irx1Kg78ZY4xZPH_sNMWIe8ifpSB_6f9HZ-JRJiVMae0b_bitAC7Wld03t6KzCdB
        use: enc
        crv: P-384
        x: f-qS0J9HcmWeU2zmDYnjCMwcsEw9ozb0_XE5y2hi2NKUJEyTgeMuWynBpexlhXbS
        y: 22-bZgbttgc4G5lXBsoVMMV5-TYg41FjJY2uGtlJp-MSfJ2agzouRjpzrCihXi7z
        alg: ECDH-ES+A256KW
```


CHANGELOG.md
## [7.48.5](https://github.com/softwaregroup-bg/ut-babelfish/compare/v7.48.4...v7.48.5) (2022-06-03)



## [7.48.4](https://github.com/softwaregroup-bg/ut-babelfish/compare/v7.48.3...v7.48.4) (2022-04-12)



## [7.48.3](https://github.com/softwaregroup-bg/ut-babelfish/compare/v7.48.2...v7.48.3) (2022-04-11)


### Bug Fixes

* use dispatch port to run ut-babelfish as part of implementation ([#2](https://github.com/softwaregroup-bg/ut-babelfish/issues/2)) ([c3076a6](https://github.com/softwaregroup-bg/ut-babelfish/commit/c3076a66c07f88ba214b666a8469c17519301fc9))



## [7.48.2](https://github.com/softwaregroup-bg/ut-babelfish/compare/v7.48.1...v7.48.2) (2021-06-23)



## [7.48.1](https://github.com/softwaregroup-bg/ut-babelfish/compare/v7.48.0...v7.48.1) (2021-06-10)



# 7.48.0 (2021-06-10)


### Features

* initial ([e380410](https://github.com/softwaregroup-bg/ut-babelfish/commit/e3804107beebeea2be6e36cf65cafb2b03154b7e))





index.test.js
require('./');
const tap = require('tap');
tap.pass('no tests yet');


test/index.test.js
const got = require('got');
const keys = require('../keys');
require('ut-run').run({
    main: [
        () => ({
            test: () => [
                require('../dispatch'),
                (...params) => require('ut-function.dispatch')({
                    namespace: ['subject', 'login'],
                    methods: {
                        'subject.object.predicate': msg => msg,
                        'login.oidc.mle'() {
                            const {sign, encrypt} = this.bus.info();
                            return {sign, encrypt};
                        }
                    }
                })(...params),
                function validation() {
                    return [
                        ({ joi }) => ({
                            'subject.object.predicate': () => ({
                                description: 'subject object predicate',
                                params: joi.object(),
                                result: joi.object(),
                                auth: 'exchange'
                            }),
                            'login.oidc.mle': () => ({
                                method: 'GET',
                                path: '/.well-known/mle',
                                auth: false
                            })
                        })
                    ];
                }
            ]
        })
    ],
    method: 'unit',
    config: {
        test: true,
        utBus: {
            serviceBus: {
                jsonrpc: {
                    ...keys
                }
            }
        },
        babelfish: {
            server: {
                port: 8091
            },
            mle: {
                routes: [{
                    path: '/rpc/subject/object/predicate',
                    auth: 'exchange'
                }]
            },
            utLog: {
                streams: {
                    udp: false
                }
            }
        }
    },
    params: {
        steps: [
            {
                name: 'subjectObjectPredicate',
                params: () => got.post('http://localhost:8091/rpc/subject/object/predicate', {
                    json: {
                        jsonrpc: '2.0',
                        id: 1,
                        method: 'subject.object.predicate',
                        params: {
                            test: true
                        }
                    }
                }).json(),
                result(msg, assert) {
                    assert.matchSnapshot(msg);
                }
            }
        ]
    }
});


index.js
const Hapi = require('@hapi/hapi');
const H2o2 = require('@hapi/h2o2');
const Wreck = require('@hapi/wreck');
const bourne = require('@hapi/bourne');
const utConfig = require('ut-config');
const Logger = require('ut-log');
const {version, name} = require('./package.json');
const cert = require('ut-bus/cert');
const merge = require('ut-function.merge');
const got = require('got');
const keys = require('./keys');
const babelfish = module.exports = async function babelfish(config) {
    const {
        server,
        proxy: {
            tls: proxyTls,
            ...proxy
        },
        log,
        utLog,
        mle
    } = utConfig.load({
        resolve: module.resolve,
        config: merge({
            implementation: 'babelfish',
            log: {
                level: 'info',
                name: 'proxy'
            },
            server: {
                host: 'localhost',
                port: 8091,
                tls: false
            },
            proxy: {
                passThrough: true,
                xforward: false,
                uri: 'http://localhost:8004',
                tls: false
            },
            utLog: {
                service: name,
                version,
                streams: {
                    stdOut: {
                        level: 'trace',
                        stream: 'process.stdout',
                        type: 'raw',
                        streamConfig: {
                            mode: 'dev'
                        }
                    },
                    udp: !process.browser && {
                        level: 'trace',
                        stream: '../udpStream',
                        streamConfig: {
                            host: 'localhost',
                            port: 30001
                        }
                    }
                }
            },
            mle: {
                routes: [
                    '/rpc/{path*}'
                ],
                ...keys
            }
        }, config)
    });

    const logger = new Logger({...utLog, streams: Object.values(utLog.streams)}).createLog(log.level, log);
    try {
        logger?.info(`Starting version ${version}`);

        logger?.info('Init client MLE keys');
        const jose = await require('ut-bus/jose')(mle);

        logger?.info(`Obtaining ${proxy.uri} MLE keys`);
        const targetKeys = await got(new URL('/rpc/login/.well-known/mle', proxy.uri).href).json();

        function createAgent() {
            if (!proxyTls) return;
            logger?.info('Loading client TLS configuration');
            const {Agent} = require('https');
            return new Agent(cert({tls: proxyTls}));
        }

        proxy.agent = createAgent();
        if (server?.tls?.key || server?.tls?.ca || server?.tls?.cert || server?.tls?.crl) {
            logger?.info('Loading server TLS configuration');
        }
        const httpServer = new Hapi.Server({...server, tls: cert(server.tls)});
        // if (this.config.capture) {
        //     await this.httpServer.register({
        //         plugin: require('ut-function.capture-hapi'),
        //         options: {name: this.config.id + '-receive', ...this.config.capture}
        //     });
        // }
        await httpServer.register([{plugin: H2o2, options: proxy}]);

        httpServer.ext('onPreResponse', (request, h) => {
            const response = request.response;
            if (response.isBoom) logger?.error(response);
            return h.continue;
        });

        httpServer.events.on('log', event => logger?.info(event));
        httpServer.events.on('start', () => logger?.info('Started', httpServer.info));

        mle.routes.concat({
            method: '*',
            path: '/{p*}',
            auth: false
        }).forEach(routeConfig => {
            const {auth, ...route} = typeof routeConfig === 'string' ? {path: routeConfig} : routeConfig;
            return httpServer.route(merge({
                method: 'POST',
                options: auth === false
                    ? {
                        auth: false,
                        payload: {
                            parse: false,
                            output: 'stream'
                        },
                        handler: {
                            proxy: {
                                uri: proxy.uri
                            }
                        }
                    }
                    : {
                        payload: {
                            parse: false,
                            output: 'data',
                            allow: ['application/json', 'application/x-www-form-urlencoded']
                        },
                        async handler(request, h) {
                            let path = '{path}';
                            let onResponse;
                            if (request.headers['content-type'] === 'application/json') {
                                const payload = bourne.parse(request.payload.toString('utf8'));

                                let keys;
                                const identityCheck = ['login.identity.check'].includes(payload.method);
                                if (identityCheck || auth === 'exchange' || payload.method.split('.').pop() === 'exchange') {
                                    keys = {
                                        mlsk: mle.sign,
                                        mlek: mle.encrypt
                                    };
                                    if (identityCheck) {
                                        path = '/rpc/login/identity/exchange';
                                        payload.method = 'login.identity.exchange';
                                    }
                                }
                                logger?.info(`Encrypting ${request.path}`);
                                payload.params = await jose.signEncrypt(payload.params, targetKeys.encrypt, keys);
                                request.payload = JSON.stringify(payload);
                                onResponse = async function DecryptVerify(e, res, request, h, settings, ttl) {
                                    const payload = await Wreck.read(res, {json: 'strict', gunzip: true});
                                    logger?.info(`Decrypting ${request.path}`);
                                    if (payload.error) payload.error = await jose.decryptVerify(payload.error, targetKeys.sign);
                                    if (payload.result) payload.result = await jose.decryptVerify(payload.result, targetKeys.sign);
                                    const response = h.response(JSON.stringify(payload));
                                    response.headers = res.headers;
                                    delete response.headers['content-length'];
                                    delete response.headers['content-encoding'];
                                    delete response.headers['transfer-encoding'];
                                    return response;
                                };
                            }
                            return h.proxy({
                                ...proxy,
                                uri: proxy.uri + path,
                                onResponse
                            });
                        }
                    }
            }, route));
        });

        logger?.info(`Starting on ${server.host}:${server.port}`);
        await httpServer.start();
        return httpServer;
    } catch (error) {
        logger?.error(error);
        throw new Error('silent');
    }
};

if (require.main === module) {
    babelfish().catch(error => {
        // eslint-disable-next-line no-console
        if (error.message !== 'silent') console.error(error);
        // eslint-disable-next-line no-process-exit
        process.exit(1);
    });
}
